### ISSUE TITLE
[MobileFormer]Converted model outputs values mismatch with original ones.

### ISSUE THREAD

**Issue by kevinz8866** [PINTO0309/onnx2tf#105]
### Issue Type

Others

### onnx2tf version number

1.4.2

### onnx version number

1.12.0

### tensorflow version number

2.10.1

### Download URL for ONNX

9e: https://drive.google.com/file/d/1vGzO9MZGX-yGz6ATm4yHVJMASZACuy2t/view?usp=share_link
9t: https://drive.google.com/file/d/1Be-a7Pmo6auyAXHChAhC1qtzkbr5EytJ/view?usp=share_link
12e: https://drive.google.com/file/d/1NR-0Fm5q_ludb5MTWgDPzRzqDFKOsCkw/view?usp=share_link
12t: https://drive.google.com/file/d/1ReuXt4gpbq6O7mpUGudaBKjb6wI43ka6/view?usp=share_link

### Parameter Replacement JSON

```json
9e: https://drive.google.com/file/d/1av1dL5sWfjlghf2-jtmws-CR-Tcv_WIP/view?usp=share_link
9t: https://drive.google.com/file/d/18qB_VOW4Xp9PvCN4G-eRRnWhh9o1fF3F/view?usp=share_link
12e: https://drive.google.com/file/d/1T0nTkb5Szf9XjaNVVBTzzu5C9dgOURLd/view?usp=share_link
12t: https://drive.google.com/file/d/1JhH8K64PDJong4qkZbzyyJDKRExE7wlJ/view?usp=share_link
```

### Description

Hi Master PINTO,

Following up on issue #103, the output values of the converted model are different from what is expected. I tried to export the pytorch model in both eval and training and both opset==9 and opset==12. For the models and parameter replacement files, the number stands for opset version, e stands for eval, and t stands for training. This issue happens in all scenarios. Please see this notebook [https://drive.google.com/file/d/1mErgTLFiYGTgUnMRDcNp3HkOkSWa96aL/view?usp=share_link](url) to replicate and for issue demonstration.

Also, I see that since you match every onnx operation with a basic tf operation, is there no way to restore those weights as trainable parameters? Thank you so much for keep digging in on this. If you need the original pytorch model or anything else I could assist on, let me know!

**Comment by kevinz8866** [comment]
By the way, these models fails to save as a keras model as well, when I run `tf.keras.models.save_model`, in addition to failure in saving in H5. The same error you found as in #103.

**Comment by kevinz8866** [comment]
If you want to use MobileFormer's pytorch repo to debug, I have a fork that fixed all the import issues in the office repo. [https://github.com/kevinz8866/MobileFormer](url)

**Comment by PINTO0309** [comment]
Although experimental, I am adding a validation function to investigate which operations of the model transformation produce errors in the output. I would eventually like to modify the tool itself to automatically check which tensors have large errors.

- [Experimental] Added the ability to validate the model final output tensor in ONNX and TensorFlow.
- This option in combination with the --output_names_to_interrupt_model_conversion option can be used to investigate which operations at which locations in the model cause errors in the output.

![image](https://user-images.githubusercontent.com/33194443/211457108-1c793beb-c801-47e1-9a0e-5882e958b6ef.png)

**Comment by PINTO0309** [comment]
- OK - `Add_116` - `onnx::MatMul_556`
![image](https://user-images.githubusercontent.com/33194443/211702313-71d05a9b-2aeb-4694-b482-6e29aed3bae8.png)
```
onnx2tf -i mobileformer.onnx -prf replace_kevinz8866.json -onimc onnx::MatMul_556 -cotoa 1e-1
```
![image](https://user-images.githubusercontent.com/33194443/211702545-6129f5c6-d378-4fa7-b1bc-f4a5a6d337df.png)

- OK - `MatMul_117` - `onnx::Add_560`
![image](https://user-images.githubusercontent.com/33194443/211702697-76dae338-27ff-42a1-a2d0-86e90f42e63d.png)
```
onnx2tf -i mobileformer.onnx -prf replace_kevinz8866.json -onimc onnx::Add_560 -cotoa 1e-1
```
![image](https://user-images.githubusercontent.com/33194443/211702858-1bd6a0b5-8ec1-4175-8b22-e7e87d2b1c03.png)

- OK - `Reshape_119` - `onnx::Transpose_569`
```
onnx2tf -i mobileformer.onnx -prf replace_kevinz8866.json -onimc onnx::Transpose_569 -cotoa 1e-1
```
![image](https://user-images.githubusercontent.com/33194443/211703068-6a6e2909-f9c0-4312-b387-684b0f768a74.png)

- NG - `MatMul_123` - `onnx::Mul_580`
```
onnx2tf -i mobileformer.onnx -prf replace_kevinz8866.json -onimc onnx::Mul_580 -cotoa 1e-1
```
![image](https://user-images.githubusercontent.com/33194443/211703300-2bc8b423-1ae0-4fff-92ef-c107577b77a2.png)
![image](https://user-images.githubusercontent.com/33194443/211703440-b4076a29-3904-4dfa-b249-574f2972baac.png)

**Comment by PINTO0309** [comment]
>Also, I see that since you match every onnx operation with a basic tf operation, is there no way to restore those weights as trainable parameters? Thank you so much for keep digging in on this. If you need the original pytorch model or anything else I could assist on, let me know!

1. First, define the model structure as a Functional model in Keras. This means that the Python code defines the model structure.
2. Extracts only the weights from the transformed model. At this time the ability to extract weights is not present in this tool, but I will try to add the feature in the future. At this time, weights can be extracted using Netron. Clicking on the floppy disk icon saves a binary file in np.ndarray format to storage.
![image](https://user-images.githubusercontent.com/33194443/211711909-13239f97-d35a-44bb-a126-42247ad09b92.png)
```python
import numpy as np
print(np.load('tensor'))

input_weights = np.load('tensor')

[[[[ 3.85814160e-02 -5.91791682e-02  4.93610911e-02]
   [-1.54751437e-02 -2.65599549e-01  2.85035670e-01]
   [-2.80502457e-02 -2.88973451e-01  2.98822671e-01]]

  [[ 4.93560582e-02 -6.98780492e-02  6.08931743e-02]
   [-1.96552109e-02 -4.20744419e-01  4.15196180e-01]
   [-1.72943212e-02 -4.09793824e-01  4.53572363e-01]]

  [[-2.43881089e-03 -2.27344222e-02  3.65414075e-04]
   [ 1.85506195e-02 -1.68236122e-01  1.95923150e-01]
   [-5.82838431e-03 -1.79103076e-01  1.61797389e-01]]]


 [[[ 1.88791680e+00  4.31701469e+00  4.70997620e+00]
```
3. Load the weights extracted in 2. into the Keras model as initializers.
```python
from tensorflow.python.keras.layers import Conv2D

Conv2D(
    filters=input_weights.shape[-1],
    kernel_size=input_weights.shape[:2],
    strides=strides,
    padding=pad_mode.lower(),
    dilation_rate=dilations,
    groups=group,
    use_bias=False,
    kernel_initializer=tf.keras.initializers.constant(input_weights),
    name=graph_node.name,
)(input_tensor)
```

**Comment by PINTO0309** [comment]
Added the ability for the tool to automatically identify operations with large model accuracy errors.

https://user-images.githubusercontent.com/33194443/211809221-d651bdcd-149f-42c1-aeec-2e43dc5aaa1e.mp4

**Comment by PINTO0309** [comment]
>By the way, these models fails to save as a keras model as well, when I run tf.keras.models.save_model, in addition to failure in saving in H5. The same error you found as in https://github.com/PINTO0309/onnx2tf/issues/103.

It is not possible to save the entire structure of the model to an h5 file, but I have added the ability to extract only the weights and save them to an h5 file. hdf5 format files are output.
```
-ow, --output_weights
  Output weights in hdf5 format.
```
![image](https://user-images.githubusercontent.com/33194443/212022784-42271bf1-4efd-493f-a1c8-0889ccd0621b.png)
![image](https://user-images.githubusercontent.com/33194443/212022798-0ec3bfb8-e9d1-4ed6-a4ce-8ce2a9bd52ff.png)

**Comment by PINTO0309** [comment]
The final outputs are now nearly identical.
- MobileFormer-e9.onnx
  https://drive.google.com/file/d/1vGzO9MZGX-yGz6ATm4yHVJMASZACuy2t/view?usp=share_link
- replace_kevinz8866.json
  ```json
  {
    "format_version": 1,
    "operations": [
      {
        "op_name": "Reshape_247",
        "param_target": "outputs",
        "param_name": "onnx::Add_753",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Reshape_418",
        "param_target": "outputs",
        "param_name": "onnx::Add_1015",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Reshape_588",
        "param_target": "outputs",
        "param_name": "onnx::Add_1275",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Reshape_759",
        "param_target": "outputs",
        "param_name": "onnx::Add_1537",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Reshape_929",
        "param_target": "outputs",
        "param_name": "onnx::Add_1797",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Reshape_1098",
        "param_target": "outputs",
        "param_name": "onnx::Add_2056",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Reshape_1269",
        "param_target": "outputs",
        "param_name": "onnx::Add_2318",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Reshape_1439",
        "param_target": "outputs",
        "param_name": "onnx::Add_2578",
        "post_process_transpose_perm": [0,2,3,1]
      },
      {
        "op_name": "Transpose_120",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_126",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_129",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_291",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_297",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_300",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_462",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_468",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_471",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_632",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_638",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_641",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_803",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_809",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_812",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_973",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_979",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_982",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_1142",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_1148",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_1151",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_1313",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_1319",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_1322",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      },
      {
        "op_name": "Transpose_1444",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [1,2,0,3]
      },
      {
        "op_name": "Softmax_1449",
        "param_target": "attributes",
        "param_name": "axis",
        "values": 3
      },
      {
        "op_name": "Transpose_1452",
        "param_target": "attributes",
        "param_name": "perm",
        "values": [2,0,1,3]
      }
    ]
  }
  ```
![image](https://user-images.githubusercontent.com/33194443/212067253-2683bc91-6274-4475-95b3-a2724aa06b1d.png)

**Comment by PINTO0309** [comment]
Because of the unpredictable transpositions in the tool's automatic transformations, I have implemented a number of enhancements to identify where the errors occur.

I believe that MobileFormer other than e9 can be converted with almost the same accuracy by changing the behavior of the tool based on the same criteria.

Once I close this issue, if you have successfully converted a model other than e9, I think many researchers and engineers would be pleased if you could pull request a sample JSON file here.
https://github.com/PINTO0309/onnx2tf/tree/main/json_samples
![image](https://user-images.githubusercontent.com/33194443/212098754-591f3851-f7c4-43dc-9af8-c875f71ac55c.png)

**Comment by PINTO0309** [comment]
@kevinz8866 Fixed a fatal bug, allowing converted models to be output to Keras (.h5).
https://github.com/PINTO0309/onnx2tf/releases/tag/1.5.30

**Comment by kevinz8866** [comment]
Hi PINTO,

Thank you so much all these updates. Sorry I was able to get back to you. I was traveling and a bit busy with the lunar new year. I will try to replicate what you had for me and yes, I will post some json files if I converted more models from onnx to keras. Thank you so much for making this package. I will keep using it and let you know if there is any issue!
